### ISSUE TITLE
configurable line colour and opacity for groupings

### ISSUE THREAD

**Issue by lvmm** [archimatetool/archi#519]
When groupings overlap with other elements it is sometimes not easy to pick up readable opacity levels. Please provide a way to control line colour and control line opacity independently of fill opacity for groupings. A grouping with no fill - just a frame, will work too.

**Comment by lvmm** [comment]
Initially I assumed this issue to be a feature request, but actually Archimate specifications explicitly state that groupings may overlap ( http://pubs.opengroup.org/architecture/archimate3-doc/chap04.html#_Toc489945986 ) which with grouping line transparency being linked to fill transparency is not easy at all, so this issue should probably be treated as a bug.

**Comment by Phillipus** [comment]
The ArchiMate specification doesn't really stipulate presentation of figures in terms of things like opacity. In fact it may be the case that a tool might not support opacity, so I regard this as a feature request.

As with all feature requests with open source software it may get implemented depending on time and resources. Contributions are welcome.

**Comment by jbsarrodie** [comment]
Hi,

>so this issue should probably be treated as a bug.

No. What ArchiMate means by "overlap" is simply that an element can be part of several Grouping.

So comming back to you original comment, I agree that it would be good to separate border color and fill color so that a user could decide to have a transparent element while still keeping a border.

I wonder what would be the best way to manage border:
* Add border opacity,
* or, add a "border visible" flag ?

As border is a bit more than border (it also includes some internal drawing for some figures like business object), I would assume "border opacity" is the best option.

**Comment by Phillipus** [comment]
Clearly such a feature should, if implemented, apply not only to the Grouping figure but to all figures. As some figures have lines inside the figure the change would require to draw those lines separately using the current opacity setting. So, this would mean re-doing the current drawing routines to separate out the lines that form the outline and those inside the figure. A few days work probably. Perhaps someone will contribute the initial work...

**Comment by Phillipus** [comment]
Something like this perhaps:

![Image1](https://user-images.githubusercontent.com/600504/63920690-50db3280-ca39-11e9-8a31-9196b9836ccc.png)

**Comment by Phillipus** [comment]
So are agreed that it is all lines in the figure that should have separate opacity?

**Comment by jbsarrodie** [comment]
> So are we agreed that it is all lines in the figure that should have separate opacity?

Yes, let's keep it simple: border opacity should be for border but also all lines inside a figure.

On question though: how to manage existing models ? Maybe something like: if border opacity is not set and fill opacity is, use the same value ?

**Comment by Phillipus** [comment]
> On question though: how to manage existing models ? Maybe something like: if border opacity is not set and fill opacity is, use the same value ?

Good question.

There has to be a default value for border opacity in the Ecore model, as there is for fill opacity, and this should be 255. However, there is no way to know if border opacity is not set because EMF returns the default value if it doesn't exist in the model. 

Therefore users will get a surprise if this is implemented.

**Comment by jbsarrodie** [comment]
Can this be implemented before loading the model in EMF. I think you have similar things happening for conversions from previous model version.

Maybe we don't even need this pre-EMF thing: we "simply" have to run a code once when an old model is opened: set borderOpacity = fillOpacity

**Comment by Phillipus** [comment]
> Maybe we don't even need this pre-EMF thing: we "simply" have to run a code once when an old model is opened: set borderOpacity = fillOpacity

Could be one of those Compatibility Handlers. I'll investigate.

**Comment by Phillipus** [comment]
The current opacity Ecore attribute is "alpha" so we have getAlpha() and setAlpha(int). We need another one. Suggestions for Ecore attribute name?

**Comment by jbsarrodie** [comment]
Beta ;-)

**Comment by Phillipus** [comment]
> Beta ;-)

(headbang emoji)

**Comment by jbsarrodie** [comment]
Joking. In fact "alpha" should be used for transparency, not opacity. So I suggest "borderOpacity" (and someday rename "alpha" to "fillOpacity"

**Comment by Phillipus** [comment]
Internally it actually is alpha. This is what is set in Draw2D and also what we set in the exchange format. I chose "opacity" in the UI because it was better than "alpha".

I would like to keep the naming consistent.

**Comment by jbsarrodie** [comment]
"borderAlpha" ?

**Comment by Phillipus** [comment]
I'm going with "lineAlpha".

**Comment by Phillipus** [comment]
> 
> 
> > Maybe we don't even need this pre-EMF thing: we "simply" have to run a code once when an old model is opened: set borderOpacity = fillOpacity
> 
> Could be one of those Compatibility Handlers. I'll investigate.

Before I investigate this let's examine the case where we don't convert line opacity to fill opacity. Do we think that not converting is unacceptable?

**Comment by jbsarrodie** [comment]
>Do we think that not converting is unacceptable?

Well, it is, but maybe providing a jArchi script for this is enough...

**Comment by Phillipus** [comment]
Actually, the Compatibility Handler for this is trivial. I'm just making sure that we agree that a conversion must take place. And I'm trying to keep these handlers to a minimum to avoid slow load times.

(A Compatibility Handler to do this adds about 100ms to load time on a model with 400 views on my machine)

**Comment by jbsarrodie** [comment]
So, IMHO, this is needed.

**Comment by Phillipus** [comment]
Fine.

**Comment by Phillipus** [comment]
With figures that display the icon, should the icon also be drawn with an alpha value? If so, which one? Line or Filll alpha? I suggest Fill alpha as it's part of the fill. This will affect figures in older models where the existing alpha has been set.

**Comment by Phillipus** [comment]
The text in the figure can also have its alpha set.

**Comment by Phillipus** [comment]
> 
> 
> I'm going with "lineAlpha".

After some experiments in the UI menus and property sections, I feel that "Fill Opacity" is unambiguous but "Line Opacity" is not as one might think of connection lines. "Border Opacity" is not accurate either because this setting affects also the internal lines of the figure. 

I think "Outline Opacity" in the UI and "outlineAlpha" for the Ecore attribute might be better.

**Comment by Phillipus** [comment]
Issues outstanding:

1. Do we agree that line/outline opacity affects not only the outer edge border but internal outlines (as seen for example in Application Component, Node, Device, Contract, etc?
1. Should text in figures inherit the fill opacity?
1. Should the icon in figures inherit the fill opacity?

**Comment by Phillipus** [comment]
Examples of figure with fill opacity of 75:

Text and Icon has fill opacity:

![Image 1](https://user-images.githubusercontent.com/600504/63963224-75acc580-ca8c-11e9-83e8-0baccfa91ac0.png)

Icon has fill opacity:

![Image 2](https://user-images.githubusercontent.com/600504/63963232-7a717980-ca8c-11e9-863b-1459bdae2649.png)

Text and Icon no opacity:

![Image 3](https://user-images.githubusercontent.com/600504/63963233-7d6c6a00-ca8c-11e9-8744-3ef15a23ad9b.png)

**Comment by jbsarrodie** [comment]
> 1. Should text in figures inherit the fill opacity?

IMHO, no. This is not related to fillColor but textColor, so if we want to implement such feature; it would be best to have a different attribute.

> 2. Should the icon in figures inherit the fill opacity?

Maybe. But in the future I would suggest a boolean attribute to show/hide icon, so this might conflict a bit.

**Comment by Phillipus** [comment]
> After some experiments in the UI menus and property sections, I feel that "Fill Opacity" is unambiguous but "Line Opacity" is not because one might think it is for connection lines. "Border Opacity" is not accurate either because this setting affects also the internal lines of the figure.
> 
> I think "Outline Opacity" in the UI and "outlineAlpha" for the Ecore attribute might be better.

The term used in drawing apps for lines is "Stroke". Would "Stroke Opacity" work, or is that too weird?

https://stackoverflow.com/questions/4125152/difference-between-stroke-and-fill

**Comment by jbsarrodie** [comment]
For sure "stroke" would be too weird. Outline is best IMHO
